File: src/bmsx/emulator/ide/lua_identifier_chain.ts
import { LuaLexer } from '../../lua/lualexer';

export function parseLuaIdentifierChain(expression: string): string[] {
	if (!expression) {
		return null;
	}
	const parts: string[] = [];
	let segmentStart = 0;
	for (let index = 0; index < expression.length; index += 1) {
		const ch = expression.charAt(index);
		if (ch !== '.' && ch !== ':') {
			continue;
		}
		const segment = expression.slice(segmentStart, index);
		if (!isValidIdentifierSegment(segment)) {
			return null;
		}
		parts.push(segment);
		segmentStart = index + 1;
	}
	const tailSegment = expression.slice(segmentStart);
	if (!isValidIdentifierSegment(tailSegment)) {
		return null;
	}
	parts.push(tailSegment);
	return parts;
}

export function resolveLuaIdentifierChainRoot(expression: string): string {
	const parts = parseLuaIdentifierChain(expression);
	if (!parts || parts.length === 0) {
		return null;
	}
	return parts[0];
}

function isValidIdentifierSegment(value: string): boolean {
	if (value.length === 0) {
		return false;
	}
	if (!LuaLexer.isIdentifierStart(value.charAt(0))) {
		return false;
	}
	for (let index = 1; index < value.length; index += 1) {
		if (!LuaLexer.isIdentifierPart(value.charAt(index))) {
			return false;
		}
	}
	return true;
}

File: packages/bmsx-console/src/res/_ignore/ide/lua_identifier_chain.lua
-- identifier_chain.lua

local identifier_chain = {}
local separators = {
	["."] = true,
	[":"] = true,
}

local function is_identifier_start(ch)
	local code = string.byte(ch)
	return code and ((code >= 65 and code <= 90) or (code >= 97 and code <= 122) or code == 95)
end

local function is_identifier_part(ch)
	local code = string.byte(ch)
	return is_identifier_start(ch) or (code and code >= 48 and code <= 57)
end

local function is_valid_identifier_segment(value)
	if #value == 0 then
		return false
	end
	if not is_identifier_start(string.sub(value, 1, 1)) then
		return false
	end
	for index = 2, #value do
		if not is_identifier_part(string.sub(value, index, index)) then
			return false
		end
	end
	return true
end

function identifier_chain.parse_identifier_chain(expression)
	if not expression or #expression == 0 then
		return nil
	end
	local parts = {}
	local segment_start = 1
	for index = 1, #expression do
		local ch = string.sub(expression, index, index)
		if separators[ch] then
			local segment = string.sub(expression, segment_start, index - 1)
			if not is_valid_identifier_segment(segment) then
				return nil
			end
			parts[#parts + 1] = segment
			segment_start = index + 1
		end
	end
	local tail_segment = string.sub(expression, segment_start)
	if not is_valid_identifier_segment(tail_segment) then
		return nil
	end
	parts[#parts + 1] = tail_segment
	return parts
end

function identifier_chain.resolve_identifier_chain_root(expression)
	local parts = identifier_chain.parse_identifier_chain(expression)
	if not parts or #parts == 0 then
		return nil
	end
	return parts[1]
end

return identifier_chain
